### PR TITLE
add support for cpuset.mems

### PR DESCRIPTION
**Pull request by hqhq**
Signed-off-by: Qiang Huang h.huangqiang@huawei.com

### CODE REVIEW

**Comment by hqhq** [comment]
It's a new version of an early one https://github.com/docker/docker/pull/9536

**Comment by crosbymichael** [comment]
LGTM

@hqhq can you please base this then it's good to merge

**Comment by ehazlett** [comment]
LGTM

**Comment by icecrime** [comment]
Pushing to `3-needs-doc-review` so we can continue progressing on this while you rebase!

File: api/client/build.go [hunk 1]
```diff
@@ -56,6 +56,7 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
 	flMemorySwap := cmd.String([]string{"-memory-swap"}, "", "Total memory (memory + swap), '-1' to disable swap")
 	flCPUShares := cmd.Int64([]string{"c", "-cpu-shares"}, 0, "CPU shares (relative weight)")
 	flCPUSetCpus := cmd.String([]string{"-cpuset-cpus"}, "", "CPUs in which to allow execution (0-3, 0,1)")
+	flCPUSetMems := cmd.String([]string{"-cpuset-mems"}, "", "MEMs in which to allow execution (0-3, 0,1)")
```
Comment (SvenDowideit) [inline api/client/build.go]: I'd add `only for NUMA systems` here - and then the man page and cli docs would say that in the cmdline flag summary too.

File: integration-cli/docker_cli_run_test.go [hunk 2]
```diff
@@ -1367,7 +1367,7 @@ func TestRunWithCpuset(t *testing.T) {
 
 	cmd := exec.Command(dockerBinary, "run", "--cpuset", "0", "busybox", "true")
 	if code, err := runCommand(cmd); err != nil || code != 0 {
-		t.Fatalf("container should run successfuly with cpuset of 0: %s", err)
+		t.Fatalf("container should run successfully with cpuset of 0: %s", err)
```
Comment (SvenDowideit) [inline integration-cli/docker_cli_run_test.go]: nice catch :)

File: runconfig/parse.go [hunk 3]
```diff
@@ -65,6 +65,7 @@ func Parse(cmd *flag.FlagSet, args []string) (*Config, *HostConfig, *flag.FlagSe
 		flWorkingDir      = cmd.String([]string{"w", "-workdir"}, "", "Working directory inside the container")
 		flCpuShares       = cmd.Int64([]string{"c", "-cpu-shares"}, 0, "CPU shares (relative weight)")
 		flCpusetCpus      = cmd.String([]string{"#-cpuset", "-cpuset-cpus"}, "", "CPUs in which to allow execution (0-3, 0,1)")
+		flCpusetMems      = cmd.String([]string{"-cpuset-mems"}, "", "MEMs in which to allow execution (0-3, 0,1)")
```
Comment (SvenDowideit) [inline runconfig/parse.go]: I'd add only for NUMA systems here - and then the man page and cli docs would say that in the cmdline flag summary too.

Comment (hqhq) [inline runconfig/parse.go]: Yeah, I wanted to add it, but we got 80 characters limit now, do you have any suggestions how we can make it more clear without break that 80 characters limit?

**Comment by hqhq** [comment]
Rebased.

File: docs/man/docker-create.1.md [hunk 4]
```diff
@@ -74,6 +75,13 @@ IMAGE [COMMAND] [ARG...]
 **--cpuset-cpus**=""
    CPUs in which to allow execution (0-3, 0,1)
 
+**--cpuset-mems**=""
+   Memory nodes (MEMs) in which to allow execution (0-3, 0,1). Only effective on NUMA systems.
+
+   If you have four memory nodes on your system(0-3), use `--cpuset-mems=0,1`
```
Comment (thaJeztah) [inline docs/man/docker-create.1.md]: needs a space after "system"

File: docs/man/docker-create.1.md [hunk 5]
```diff
@@ -74,6 +75,13 @@ IMAGE [COMMAND] [ARG...]
 **--cpuset-cpus**=""
    CPUs in which to allow execution (0-3, 0,1)
 
+**--cpuset-mems**=""
+   Memory nodes (MEMs) in which to allow execution (0-3, 0,1). Only effective on NUMA systems.
```
Comment (thaJeztah) [inline docs/man/docker-create.1.md]: s/in which/on which/ (I think)

File: docs/man/docker-create.1.md [hunk 6]
```diff
@@ -74,6 +75,13 @@ IMAGE [COMMAND] [ARG...]
...
+
+   If you have four memory nodes on your system(0-3), use `--cpuset-mems=0,1`
+then processes in your Docker container will only use memory from the first
+two memory nodes.
```
Comment (thaJeztah) [inline docs/man/docker-create.1.md]: This sentence doesn't "run" properly. Thinking something like (but I'm no doc-writer);

On a system with four memory nodes (0 - 3), setting `--cpusets-mems=0,1` will
restrict the Docker container to only use memory from the first two memory nodes.

File: docs/sources/reference/api/docker_remote_api_v1.19.md [hunk 7]
```diff
@@ -191,6 +192,7 @@ Json Parameters:
       (ie. the relative weight vs other containers).
 -   **Cpuset** - The same as CpusetCpus, but deprecated, please don't use.
 -   **CpusetCpus** - String value containing the cgroups CpusetCpus to use.
+-   **CpusetMems** - Memory nodes (MEMs) in which to allow execution (0-3, 0,1). Only effective on NUMA systems.
```
Comment (thaJeztah) [inline docs/sources/reference/api/docker_remote_api_v1.19.md]: same here: s/in/on/

File: docs/sources/reference/run.md [hunk 8]
```diff
@@ -599,6 +600,19 @@ This means processes in container can be executed on cpu 1 and cpu 3.
...
 This means processes in container can be executed on cpu 0, cpu 1 and cpu 2.
 
+We can set mems in which to allow execution for containers. Only effective
+on NUMA systems.
+
+Examples:
+
+    $ docker run -ti --cpuset-mems="1,3" ubuntu:14.04 /bin/bash
+
+This means processes in container can be executed on memory node 1 and memory node 3.
```
Comment (thaJeztah) [inline docs/sources/reference/run.md]: "This example restricts the processes in the container to only use memory from memory nodes 1 and 3."

Are the quotes surrounding "1,3" required or optional? Earlier examples didn't use quotes.

File: docs/man/docker-create.1.md (hunk 5 above)
Comment (hqhq) [inline docs/man/docker-create.1.md]: Let's leave it the same as CPUs, that's not a big deal right?

File: docs/man/docker-create.1.md (hunk 4 above)
Comment (hqhq) [inline docs/man/docker-create.1.md]: I'll take it, thanks.

File: docs/sources/reference/run.md (hunk 8 above)
Comment (hqhq) [inline docs/sources/reference/run.md]: It's better, thanks.

**Comment by hqhq** [comment]
@thaJeztah Thanks for your review, since docs in this PR are almost the same as it in https://github.com/docker/docker/pull/9536, which are already been fully discussed, and got two docs LGTM from @SvenDowideit and @fredlf .
So for some A or B choice and sentence change I just leave it as it was. Hope you are OK with that :)

File: docs/sources/reference/run.md (hunk 8 above)
Comment (hqhq) [inline docs/sources/reference/run.md]: Quotes are optional, it's same as all other options.

**Comment by hqhq** [comment]
Updated, thanks @thaJeztah

**Comment by thaJeztah** [comment]
Thanks!

File: docs/man/docker-create.1.md (hunk 5 above)
Comment (thaJeztah) [inline docs/man/docker-create.1.md]: Ah, I see that now. Will leave it to the docs maintainers :)

**Comment by moxiegirl** [comment]
LGTM

**Comment by moxiegirl** [comment]
ping @jamtur01 @fredlf @SvenDowideit

**Comment by jamtur01** [comment]
LGTM

**Comment by cpuguy83** [comment]
LGTM

**Pull request merged**